### PR TITLE
Extend Browser API to check for custom text location

### PR DESCRIPTION
**Pull request by amartya4256**
The contribution provides an API to check if the location is the location where custom texts are loaded in a browser. By construct, "about:blank" is used for that purpose. This API can be adapted by the clients instead of explicitly defining the string and checking.

Contributes to #213

### CODE REVIEW

**Review by HeikoKlare — COMMENTED** [review]
Seeing the implementation, I am wondering whether it would also be sufficient to just add a `showsCustomText()` method instead of the `isLocationForCustomText(String)` one. Then the functionality would be completely independent from the location and the browser handle the state of whether it shows custom text or some URI-based page completely internally and even better hide it from consumers.
@amartya4256 Can you check whether that's possible?

Edit: Please also validate against the WebKit implementation. It seems to already have the about:blank page encoded as kind of a base URI internally. It also has some comments about using about:blank to be consistent with IE, so maybe there is also some further chance to align. I.e., a comment states: 
https://github.com/eclipse-platform/eclipse.platform.swt/blob/97cf720f1a64c04de12998b8a50cf57c7189c5f8/bundles/org.eclipse.swt/Eclipse%20SWT%20WebKit/cocoa/org/eclipse/swt/browser/WebKit.java#L634-L637

**Comment by amartya4256** [comment]
> Seeing the implementation, I am wondering whether it would also be sufficient to just add a `showsCustomText()` method instead of the `isLocationForCustomText(String)` one. Then the functionality would be completely independent from the location and the browser handle the state of whether it shows custom text or some URI-based page completely internally and even better hide it from consumers. @amartya4256 Can you check whether that's possible?
> 
> Edit: Please also validate against the WebKit implementation. It seems to already have the about:blank page encoded as kind of a base URI internally. It also has some comments about using about:blank to be consistent with IE, so maybe there is also some further chance to align. I.e., a comment states:
> 
> https://github.com/eclipse-platform/eclipse.platform.swt/blob/97cf720f1a64c04de12998b8a50cf57c7189c5f8/bundles/org.eclipse.swt/Eclipse%20SWT%20WebKit/cocoa/org/eclipse/swt/browser/WebKit.java#L634-L637

using showsCustomText() won't be possible everywhere since there are Url comparison before the location of the browser has changed. For example in location listener in the changing method which executes before the location is actually changed. Another option was to cache the state type on setText but we don't know where we land with back and forward calls so it cannot be fully implemented. Moreover, I find the current method fine to accept a location as an argument. If you want to check something explictly, that can be passed, otherwise browser.getUrl() can be passed.

File: bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/Browser.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/Browser.java:1193]: Would be good to extend the documentation a bit on what "setting custom text" means. Somethin like:
```suggestion
 * Checks if the location supplied is the location for setting custom text in the browser. This means, when passing the browser URL to this method after a custom text has been set via {@link #setText(String)} will yield true as long as no further navigation to some other location is performed.
```

File: bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/Browser.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/Browser.java:1200]: I propose to have a null check for `location` here. Either specify what happen when passing null (such as always returning false for a null value), or explicitly specifying in the Javadoc that null is not allowed and additionally explicitly checking the parameter to have an early failure:
```suggestion
	return URI.create(Objects.requireNonNull(location)).equals(webBrowser.getBaseUri());
```

File: bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/WebBrowser.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/WebBrowser.java:729]: Can this be package protected, since the only intended consumer (`Browser´) is in the same package? Please also improve the name according to field name (see comment on the field name for that).

In addition, please check whether this method can be used by the IE and WebKit implementations, which currently have their own `ABOUT_BLANK` field, which might (at least at some places) be reasonably replaced by this method.

File: bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/WebBrowser.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/WebBrowser.java:40]: Please change this to a more self-explaining name, such as `URI_FOR_CUSTOM_TEXT_PAGE`.

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (HeikoKlare) [inline tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:530]: The assertion message is unrelated to what the assertion actually checks. I propose to make two assertions out of this:
1. one assertion checking that `LocationListener.changing()` has fired
2. one assertion checking the return value of the new method.

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (HeikoKlare) [inline tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:524]: Please extend the tests to not only cover the happy path. I consider at least the following scenarios relevant (maybe there are even more):
- check that location after initialization is not the location for custom text
- check that location after moving to some "actual" URL is not considered a location for custom text
- check that when calling `setText()` first and then navigating to some other page via `setUrl()` results in the location not being considered the location for custom text
- check that when calling `setText()` twice and using the browser Url after the first call of `setText()` after the second call of `setText()` to check whether it is a location for custom text still yields true (this is for the case that some asynchronous navigation is performed while some consumer still wants to check the previous URL)
- validate what happens if calling that method after the browser has been disposed

**Review by HeikoKlare — COMMENTED** [review]
The API extension in general looks good to me. I have some comments on naming, documentation and testing.

Please also adapt the commit message header (and PR name) to be more self-explaining. I.e., instead of "Added API for checking if BASE_URI" use something like "Extend Browser API to check for custom text location"

File: bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/Browser.java
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/Browser.java:1193]: Done

File: bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/Browser.java
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/Browser.java:1200]: Done

File: bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/WebBrowser.java
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/WebBrowser.java:729]: Made it package protected but I think refactoring of the browser specific usage can be done in a follow up PR?

File: bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/WebBrowser.java
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/WebBrowser.java:40]: Done

File: bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/Browser.java
Comment (HeikoKlare) [reply bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/Browser.java]: Nothing changed in the specification, only the null check was added to the code.

File: bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/WebBrowser.java
Comment (HeikoKlare) [reply bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/WebBrowser.java]: That's also fine. But then it should be done directly after merging this change.

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (HeikoKlare) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:524]: I see, So probably the first proposed test case does not make any sense. No need to add a different kind of test for that. This should only be about testing the contract of the new `isLocationForCustomText()` method.

File: bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/Browser.java
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/Browser.java]: Added in the parameter specification

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (amartya4256) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:577]: Added the tests except for the 4th point. I don't get it. Can you explain a bit?

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (HeikoKlare) [inline tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:548]: Since this operation is asynchronous, you should wait for it to complete before calling `setUrl()`. Currently there is a race condition, as `setText()` may be completely processed before `setUrl()` is executed or not, thus potentially resulting in flaky test results in case of a regression.

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (HeikoKlare) [inline tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:526]: This test is unrelated to the PR, isn't it? There should be a test for behavior of `isLocationForCustomURL()` after disposal instead/in addition.

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (HeikoKlare) [inline tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:553]: Please make two assertions out of this, to identify which of the calls leads to a failure. The URI comparison can even be replaced with an `assertEquals` to get better output in case of a failure.

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (HeikoKlare) [inline tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:530]: Since all the added tests are dedicated to test behavior of `isLocationForCustomText()`, it would be good to have the test names always start with the tested method and append the use case to that. E.g., `test_isLocationForCustomText_falseAfterSetUrl`.

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (HeikoKlare) [inline tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:541]: Would be nice to give these guard variables a more self-explaining name like `locationChanged`.

**Review by HeikoKlare — COMMENTED** [review]
Thank you for adding the tests! In general, they look fine. My comments are mostly nitpicky style proposals, but some are also on potential race conditions. Anyway, all of them should be easy to resolve.

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (amartya4256) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:532]: Its for debugging.

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (amartya4256) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:538]: Done

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (amartya4256) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:562]: Done

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (amartya4256) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java]: Done

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (amartya4256) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java]: Done

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (amartya4256) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java]: Done

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (amartya4256) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java]: Done

File: bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/Browser.java
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT Browser/common/org/eclipse/swt/browser/Browser.java]: Done

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (amartya4256) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java]: Done

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (HeikoKlare) [inline tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:542]: Just wondering: it does not seem to matter when (or even it at all) the shell is opened. Still is seems weird that it is opened after all operations have already been executed. Maybe the shell should be opened at the beginning of the test to avoid confusion (at least it should be done at a consistent point in time throughout the tests)? Or is there a reason for doing it here?

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (HeikoKlare) [inline tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:568]: I still do not understand the idea of this logging. The `testLogAppend()` seems to be used for writing to a local log that is evaluated only at very specifiy places (e.g. `validateTitleChanged()`). Is the value ever used in the related test cases? Otherwise, e.g., if this is only for local debugging purposes by manually inspecting the written variable in a debugger, you may also just set a breakpoint here and take a look at the original variable (`pluginPath`).
Don't get me wrong: writing to this log does not really hurt, but it leaves future maintainers future with the question what the intent of executing this operation was (as it not obviously useful).

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (HeikoKlare) [inline tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:1127]: This assertion seems misplaced. The test is about validating the behavior of `setText()`, now it might fail because `isLocationForCustomText()` is buggy. So if this is a relevant scenario that is not already covered by the `test_isLocationForCustomText_*()` method, please add an according one instead of mixing up concerns here.

**Review by HeikoKlare — COMMENTED** [review]
Thanks again for the recent update! I have some minor questions/comments left.

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (amartya4256) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:1127]: It has been tested for setText in many isLocationForCustomText tests. So, I'll just remove this here.

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (amartya4256) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:568]: I just used one of the fragments which was generating the URL in one of the tests. To me, it has no use. So I can also remove it.

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java
Comment (amartya4256) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java:542]: It doesnt matter at all haha. Te tests even work without opening the shell. It's just for the visibility's sake but I think, if we don't open the shell, the tests run faster. Also we don't have those blinking windows on the monitor for the time the tests are running.

Comment (amartya4256) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_browser_Browser.java]: Removed the shell.open calls from new tests

**Review by HeikoKlare — APPROVED** [review]
From my side, this is ready to be merged once master branch is open again. Thank you!

**Pull request merged**
